drivers/flash/flash_stm32h7x.c

```diff
@@ -137,6 +159,74 @@ void flash_stm32_set_rdp_level(const struct device *dev, uint8_t level)
 }
 #endif /* CONFIG_FLASH_STM32_READOUT_PROTECTION */
 
+#if defined(CONFIG_FLASH_STM32_WRITE_PROTECT)
+
+#define WP_MSK FLASH_WPSN_WRPSN_Msk
+#define WP_POS FLASH_WPSN_WRPSN_Pos
+
+int flash_stm32_update_wp_sectors(const struct device *dev, uint64_t changed_sectors,
+				  uint64_t protected_sectors)
+{
+	uint64_t bank_mask = WP_MSK >> WP_POS;
+	const uint32_t sectors_per_bank = __builtin_popcount(WP_MSK);
+	uint64_t sectors_mask = 0;
+	uint32_t protected_sectors_reg;
+	uint32_t changed_sectors_reg;
+	int ret;
+	bool commit = false;
+
+	for (int i = 0; i < NUMBER_OF_BANKS; i++) {
+		sectors_mask |= bank_mask << (sectors_per_bank * i);
+	}
+
+	if ((changed_sectors & sectors_mask) != changed_sectors) {
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < NUMBER_OF_BANKS; i++) {
+		/* Prepare protected and changed masks per bank. */
+		protected_sectors_reg = (protected_sectors >> sectors_per_bank * i) & bank_mask;
+		changed_sectors_reg = (changed_sectors >> sectors_per_bank * i) & bank_mask;
+
+		if (changed_sectors == 0) {
+			continue;
+		}
+		changed_sectors_reg <<= WP_POS;
+		protected_sectors_reg <<= WP_POS;
+		/* Sector is protected when bit == 0. Flip protected_sectors bits */
+		protected_sectors_reg = ~protected_sectors_reg;
+
+		/* Commit option bytes changes, when all registers are prepared. */
+		if (i == (NUMBER_OF_BANKS - 1)) {
+			commit = true;
+		}
+		ret = write_optwp(dev, changed_sectors_reg, protected_sectors_reg, i, commit);
+		if (ret != 0) {
+			return ret;
+		}
```

Comment:
I think, error handling here is not that simple as you might expect. What would happen if the `write_optwp()` function fails for the second bank? What would happen if we fail to update the first bank, but the second would succeed? Even if we commit option bytes outside of the loop as I suggest in the other comment, returning error here will basically skip commit.

Reply:
Am I understanding correctly that you think it would be good to commit changes, even if call `write_optwp(dev, changed_sectors_reg, protected_sectors_reg, i, false);` fails for one of the banks? (not current version)

Reply:
You are right.